src/Functions/serial.cpp

```diff
@@ -0,0 +1,171 @@
+
+namespace DB {
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+class FunctionSerial : public IFunction
+{
+private:
+    mutable zkutil::ZooKeeperPtr zk{nullptr};
+    ContextPtr context;
+
+public:
+    static constexpr auto name = "serial";
+
+    explicit FunctionSerial(ContextPtr ctx) : context(ctx)
+    {
+        if (ctx->hasZooKeeper()) {
+            zk = ctx->getZooKeeper();
+        }
+    }
+
+    static FunctionPtr create(ContextPtr context)
+    {
+        return std::make_shared<FunctionSerial>(std::move(context));
+    }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool isStateful() const override { return true; }
+    bool isDeterministic() const override { return false; }
+    bool isDeterministicInScopeOfQuery() const override { return false; }
+    bool isSuitableForConstantFolding() const override { return false; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool useDefaultImplementationForNothing() const override { return false; }
+    bool canBeExecutedOnDefaultArguments() const override { return false; }
+    bool isInjective(const ColumnsWithTypeAndName & /*sample_columns*/) const override { return true; }
+    bool hasInformationAboutMonotonicity() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() != 1)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Number of arguments for function {} doesn't match: passed {}, should be 1.",
+                getName(), arguments.size());
+        if (!isStringOrFixedString(arguments[0])) {
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Type of argument for function {} doesn't match: passed {}, should be string",
+                getName(), arguments[0]->getName());
+        }
+
+        return std::make_shared<DataTypeInt64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        auto col_res = ColumnVector<Int64>::create();
+        typename ColumnVector<Int64>::Container & vec_to = col_res->getData();
+        size_t size = input_rows_count;
+        LOG_INFO(getLogger("Serial Function"), "Size = {}", size);
+        vec_to.resize(size);
+
+        const auto & serial_path = "/serials/" + arguments[0].column->getDataAt(0).toString();
+
+        // if serial name used first time
+        zk->createAncestors(serial_path);
+        zk->createIfNotExists(serial_path, "");
+
+        Int64 counter;
+
+        if (zk != nullptr) {
+            // Get Lock in ZooKeeper
```

Comment:
It is much better to use compare-and-swap operation in a form of a `multi` with version check.

Reply:
Didn't get how to use `multi` for CAS. Now using 2 operations: `get` and `trySet` with version check in loop